crates/node-maintainer/src/maintainer.rs

```diff
@@ -120,113 +131,177 @@ impl NodeMaintainer {
     }
 
     async fn run_resolver(&mut self) -> Result<(), NodeMaintainerError> {
-        let mut package_groups = Vec::new();
-        let mut q = VecDeque::new();
-        q.push_back(self.graph.root);
-        // Start iterating over the queue. We'll be adding things to it as we find them.
-        while !q.is_empty() {
-            while let Some(node_idx) = q.pop_front() {
-                let mut names = HashSet::new();
-                let mut packages = Vec::new();
-                let manifest = self.graph[node_idx]
-                    .package
-                    .corgi_metadata()
-                    .await?
-                    .manifest;
-                // Grab all the deps from the current package and fire off a
-                // lookup. These will be resolved concurrently.
-                for ((name, spec), dep_type) in self.package_deps(node_idx, &manifest) {
-                    // `dependencies` > `optionalDependencies` ->
-                    // `peerDependencies` -> `devDependencies` (if we're looking
-                    // at root)
-                    let name = UniCase::new(name.clone());
-                    if !names.contains(&name) {
-                        let requested = format!("{name}@{spec}").parse()?;
+        let idx_queue = Arc::new(Mutex::new(VecDeque::from([self.graph.root])));
+
+        // Number of nodes in processing. When this drops to zero we close
+        // both idx_receiver and dep_receiver so that streams end.
+        let in_flight_count = Arc::new(atomic::AtomicUsize::new(1));
+
+        let nassun = Arc::new(&self.nassun);
+        let graph = Arc::new(Mutex::new(&mut self.graph));
+
+        let (idx_sender, idx_receiver) = mpsc::unbounded::<()>();
+        idx_sender.unbounded_send(())?;
+
+        let (dep_sender, dep_receiver) = mpsc::unbounded::<NodeDependency>();
+
+        let dep_receiver = dep_receiver.map(|node_dep| {
+            let idx_queue = idx_queue.clone();
+            let in_flight_count = in_flight_count.clone();
+            let graph = graph.clone();
+            let nassun = nassun.clone();
+            let idx_sender = idx_sender.clone();
+            let dep_sender = dep_sender.clone();
+
+            async move {
+                let package = nassun.resolve(&node_dep.name).await?;
```

Comment:
Why only the name here? Don't you want `format!("{name}@{spec}")`, like it was before? Otherwise, it'll just look up and resolve whatever `latest` is for that package when fetching the corgi.

Reply:
Good catch! I guess I lost it in one of refactors. Thanks!

Reply:
Oh wow, it is slower than main now, it seems...